x/multisig/decorator.go

```diff
@@ -33,62 +33,49 @@ func (d Decorator) Check(ctx weave.Context, store weave.KVStore, tx weave.Tx, ne
 // Deliver enforces multisig contract before calling down the stack
 func (d Decorator) Deliver(ctx weave.Context, store weave.KVStore, tx weave.Tx, next weave.Deliverer) (weave.DeliverResult, error) {
 	var res weave.DeliverResult
-	newCtx, err := d.withMultisig(ctx, store, tx)
+	newCtx, err := d.authMultisig(ctx, store, tx)
 	if err != nil {
 		return res, err
 	}
 
 	return next.Deliver(newCtx, store, tx)
 }
 
-func (d Decorator) withMultisig(ctx weave.Context, store weave.KVStore, tx weave.Tx) (weave.Context, error) {
-	if multisigContract, ok := tx.(MultiSigTx); ok {
-		ids := multisigContract.GetMultisig()
-		for _, contractID := range ids {
-			if contractID == nil {
-				return ctx, nil
-			}
+func (d Decorator) authMultisig(ctx weave.Context, store weave.KVStore, tx weave.Tx) (weave.Context, error) {
+	multisigContract, ok := tx.(MultiSigTx)
+	if !ok {
+		return ctx, nil
+	}
 
-			// check if we already have it
-			if d.auth.HasAddress(ctx, MultiSigCondition(contractID).Address()) {
-				return ctx, nil
-			}
+	ids := multisigContract.GetMultisig()
+	for _, contractID := range ids {
+		if contractID == nil {
+			return ctx, nil
+		}
 
-			// load contract
-			contract, err := d.getContract(store, contractID)
-			if err != nil {
-				return ctx, err
-			}
+		// If already authenticated it does not matter if multisig can
+		// authenticate as well. Any authentication method is enough.
+		if d.auth.HasAddress(ctx, MultiSigCondition(contractID).Address()) {
+			return ctx, nil
+		}
 
-			// collect all sigs
-			sigs := make([]weave.Address, len(contract.Sigs))
-			for i, sig := range contract.Sigs {
-				sigs[i] = sig
-			}
+		contract, err := d.bucket.GetContract(store, contractID)
+		if err != nil {
+			return ctx, err
+		}
 
-			// check sigs (can be sig or multisig)
-			authenticated := x.HasNAddresses(ctx, d.auth, sigs, int(contract.ActivationThreshold))
-			if !authenticated {
-				return ctx, errors.Wrapf(errors.ErrUnauthorized, "contract=%X", contractID)
+		var power Weight
+		for _, p := range contract.Participants {
```

Comment:
Note this is expensive based on number of *possible* participant, not even signatures,
make sure there is a reasonable limit on number of participants, or at least a cost based on that. (see comment above)

Reply:
The limit is 100 as specified in the issue :man_scientist:

Reply:
note to self: rtfm